_posts/on-viewmodel-composition-and-ui-composition.md

```diff
@@ -0,0 +1,152 @@
+
+```html
+<h2>Available products</h2>
+<ul>
+  @foreach (dynamic availableProduct in Model)
+  {
+    <li>
+    <a asp-controller="Products"
+         asp-action="Details"
+         asp-route-id="@availableProduct.Id">@availableProduct.ProductName</a>,
+         @await Component.InvokeAsync("Sales.ViewComponents.AvailableProductPrice", availableProduct)
+    </li>
+  }
+</ul>
+```
+
+Similarly, the component doesn't know how to render prices; it delegates the responsibility to a component from Sales.
+
+More sophisticated samples involve cross-component communication. For example, a menu component, when clicked, can cause another one on the page, e.g., a gauge to show a different set of data.
+
+We can push even further. Instead of having a known markup skeleton, we could be using a mechanism to discover components at runtime and inject them into the user interface. For example, it's what the data template mechanism does in WPF (Windows Presentation Foundation).
+
+I'm barely scratching the surface here; this post's goal is to discuss the relationship between UI and ViewModel Composition and not detail the technologies involved. Suppose you want to dive into UI Composition and micro frontends; in that case, a good starting point is the [OpenComponents framework](https://opencomponents.github.io), my friend [Mattia](https://twitter.com/mattiaerre) that is one of the contributors, and the folks at OpenTable.
+
+## Fun, fun, fun. So, what's the problem?
+
+The presented solution works very well, regardless of the technology we select to implement it. If the more than legit goal is to componentize the user interface, UI Composition is a good solution. However, not all that glitters is gold. Let's take into account the following pseudo user interface markup:
+
+```xml
+<section>
+   <product-name-and-description />
+   <product-price />
+<section>
+```
+
+Each one of the `<product-* />` elements is an autonomous component.
+
+### What's the data source?
+
+Let's imagine the presented user interface lives on a web page; given that it represents a product, a URL probably identifies it, e.g. `/products/123`. At runtime, the browser loads the page. The UI Composition framework of choice loads the data using URL data as part of the resource identifier. Each component can extract from the loaded product the portion they need to display data.
+
+If the system is a distributed system, it's legit to expect that product name and description come from the Catalog service and price comes from the Sales service. At this point, the UI Composition framework cannot make any assumption and cannot load data on behalf of the components as it doesn't know where to load the data. Components, being autonomous, can load data by themselves.
+
+The consequence is that instead of one HTTP call to load a product, we need two of them, each one issued by a component, going to different services. Fair enough, it's the cost of autonomy, and it comes with significant advantages.
```

Comment:
```suggestion
The consequence is that instead of one HTTP call to load a product, we need two of them, each one issued by a component going to different services. Fair enough, it's the cost of autonomy, and it comes with significant advantages.
```